### PR TITLE
1903 save load algorithm task data

### PR DESCRIPTION
**Pull request by ramtob**

### CODE REVIEW

File: src/app/@ansyn/ansyn/app-providers/app-providers.module.ts [hunk 1]
```diff
@@ -53,7 +53,7 @@ import { IdahoSourceProvider } from './overlay-source-providers/idaho-source-pro
 
 		{ provide: ProjectionService, useClass: OpenLayersProjectionService },
 
-		{ provide: AlgorithmsService, useClass: AlgorithmsDefaultService }
+		{ provide: TasksRemoteService, useClass: TasksRemoteDefaultService }
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/ansyn/app-providers/app-providers.module.ts]: done

File: src/app/@ansyn/menu-items/algorithms/components/tasks-table/tasks-table.component.ts [hunk 2]
```diff
@@ -1,16 +1,115 @@
...
+export class TasksTableComponent implements OnInit, OnDestroy {
+	@ViewChild('tbodyElement') tbodyElement: ElementRef;
+
+	taskState$: Observable<ITasksState> = this.store$.select(tasksStateSelector);
+
+	ids$: Observable<string[] | number[]> = this.store$.select(selectTasksIds);
+	entities$: Observable<Dictionary<AlgorithmTaskPreview>> = this.store$.select(selectTaskEntities);
+
+	@AutoSubscription
+	selectedTaskId$: Observable<string> = this.taskState$.pipe(
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/components/tasks-table/tasks-table.component.ts]: done

File: src/app/@ansyn/menu-items/algorithms/services/tasks-remote-default.service.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import { TasksRemoteService } from './tasks-remote.service';
+import { AlgorithmTask } from '../models/tasks.model';
+
+@Injectable()
+export class TasksRemoteDefaultService implements TasksRemoteService {
+
+	constructor() {
+	}
+
+	runTask(task: AlgorithmTask) {
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks-remote-default.service.ts]: done

File: src/app/@ansyn/menu-items/algorithms/services/tasks-remote.service.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { AlgorithmTask } from '../models/tasks.model';
+
+export abstract class TasksRemoteService {
+	abstract runTask(task: AlgorithmTask);
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks-remote.service.ts]: done

File: src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		protected storageService: StorageService,
+		public errorHandlerService: ErrorHandlerService
+	) {
+	}
+
+	loadTasks(tasksOffset: number = 0): Observable<any> {
+		return this.storageService.getPage<AlgorithmTaskPreview>(this.config.schema, tasksOffset, this.config.paginationLimit)
+			.pipe(
+				map(previews => previews.map(preview => this.parseTaskPreview(preview))),
+				catchError(err => this.errorHandlerService.httpErrorHandle(err, 'Failed to load tasks'))
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts]: done

[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts]: done

File: src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+			preview: this.getEntityPreview(task),
+			data: this.getEntityData(task.state)
+		};
+	}
+
+	createTask(selectedTask: AlgorithmTask): Observable<AlgorithmTask> {
+		return this.storageService.create(this.config.schema, this.convertToStoredEntity(selectedTask))
+			.pipe<any>(
+				map(_ => selectedTask),
+				catchError(err => this.errorHandlerService.httpErrorHandle(err, 'Failed to create task'))
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts]: done

File: src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+			);
+	}
+
+	removeTask(selectedTaskId: string): Observable<any> {
+		return this.storageService.delete(this.config.schema, selectedTaskId).pipe(
```
[the remark this answers is not in the document]
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/services/tasks.service.ts]: done

File: src/app/@ansyn/ansyn/ansyn.module.ts [hunk 8]
```diff
@@ -42,7 +42,7 @@ import { UploadsModule } from '@ansyn/menu-items';
 		FiltersModule,
 		LayersManagerModule,
 		ToolsModule,
-		TasksModule,
+		TasksModule.provideRemote(TasksRemoteDefaultService),
```
Comment (yairtawil) [inline src/app/@ansyn/ansyn/ansyn.module.ts]: cool

File: src/app/@ansyn/core/services/error-handler.service.ts [hunk 9]
```diff
@@ -11,14 +11,18 @@ export class ErrorHandlerService {
...
 		this.loggerService.error(errMsg);
 		this.store.dispatch(new SetToastMessageAction({
 			toastText: toastMessage || 'Connection Problem',
 			showWarningIcon: true
 		}));
-		return throwError(errMsg);
+		if (typeof returnValue === 'undefined') {
+			return throwError(errMsg);
+		} else {
+			return of(returnValue)
```
Comment (yairtawil) [inline src/app/@ansyn/core/services/error-handler.service.ts]: what about EMPTY ?

File: src/app/@ansyn/menu-items/algorithms/components/remove-task-modal/remove-task-modal.component.html [hunk 10]
```diff
@@ -1,10 +1,10 @@
 <div class="remove-task-modal">
 	<form #f="ngForm"
-		  (ngSubmit)="onSubmit(false)">
+		  (ngSubmit)="onSubmit(true)">
 		{{message}}
 		<div class="bottom-btns">
 			<button type="button" (click)="onSubmit(false)">Cancel</button>
-			<button type="button" (click)="onSubmit(true)">OK</button>
+			<button type="submit">OK</button>
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/algorithms/components/remove-task-modal/remove-task-modal.component.html]: GREAT

File: src/app/@ansyn/menu-items/algorithms/services/tasks-remote-default.service.ts [hunk 11]
```diff
@@ -1,14 +1,16 @@
...
 @Injectable()
 export class TasksRemoteDefaultService implements TasksRemoteService {
 
 	constructor() {
 	}
 
-	runTask(task: AlgorithmTask) {
+	runTask(task: AlgorithmTask): Observable<any> {
 		console.log('running task', task);
+		return of(true);
```
Comment (yairtawil) [inline src/app/@ansyn/menu-items/algorithms/services/tasks-remote-default.service.ts]: Wonderful

File: src/app/@ansyn/ansyn/ansyn.module.ts (hunk 8 above)
Comment (ramtob) [reply src/app/@ansyn/ansyn/ansyn.module.ts]: 👍

File: src/app/@ansyn/menu-items/algorithms/components/remove-task-modal/remove-task-modal.component.html (hunk 10 above)
Comment (ramtob) [reply src/app/@ansyn/menu-items/algorithms/components/remove-task-modal/remove-task-modal.component.html]: 🥇

File: src/app/@ansyn/core/services/error-handler.service.ts (hunk 9 above)
Comment (ramtob) [reply src/app/@ansyn/core/services/error-handler.service.ts]: we'll see

**Pull request merged**